### ISSUE TITLE
`m17n:sa:itrans` move the input cursor to the head

### ISSUE THREAD

**Issue by itt533** [ibus/ibus#2536]
Please fill in the following items if you don't know the root cause.

Which distribution and version?: 
Slackware 15.0, 64bit

Which desktop environment and version?: 
Xfce 4.16

Which session type?: 
X11

Which application and version?: 
anything that take text input. Here Mousepad 0.5.8 is taken

IBus version?: 
1.5.25

Issue description:
when issuing some text in a language script containing ligatures like Indic  scripts with Ibus, Ibus displays the text gradually as it is typed, in a temporary field that the program in which it is typed (here Mousepad) does not see. That temporary text is then validated with either a key stroke that makes it complete with regards to the corresponding language grammar, or a space stroke or an arrow. By looking carefully at the text during the moment Ibus validates and  passes that text to Mousepad, one can see that Ibus makes that text blinking once and the cursor looks like going backward by one text unit and move forward. When typing for just about 3~4 lines it can really become annoying. It's like giving a kick in the screen at every syllable. Off course it means that the code in Ibus is not proper.

Emacs has its own foreign input system. I have made a screencast comparing Ibus and Emacs. One could easily see that typing in Emacs looks perfectly smooth.

here is 

Steps to reproduce:
1. enable Ibus, and add 'sanskrit - itrans' or 'telugu - itrans' if not yet there
2. open Mousepad, or Kate, or whatever GUI text editor (even Firefox is a good test bed)
3. select that script mentioned in step 1
4. type 'bhA `space`' it should output 'भा' if 'sanskrit - itrans', and 'భా' if 'telugu - itrans'
5. repeat the step 4 many times by hitting `space` in between. It is clear that the text blinks once and the cursor disappears and reappears. Compare with Emacs foreign input system. (`C-x` `C-m` `C-\` prompts for an input method, type 'deva' + `tab` `tab` and choose 'devanagari-itrans' or type 'te-itans'  `tab` and `Enter` for telugu)

here is the screencast: http://0x0.st/Hjtu.mkv

Can you reproduce your problem when you restart ibus-daemon? (yes / no): 
(Run `ibus exit` and `ibus-daemon --xim &`)
YES

Do you see any errors when you run ibus-daemon with the verbose option?:
(Run `ibus-daemon --xim --verbose &` and look at the output when you encounter your problem.)
NO

Can you reproduce your problem with a new user account instead of the current your account? (yes / no): 
YES

**Comment by mike-fabian** [comment]
>  I don't know which IBus component includes the "sanskrit - itrans" engine. 

This one:

```
$ ibus list-engine | grep sa-itrans
  m17n:sa:itrans - sa-itrans (m17n)
```

**Comment by mike-fabian** [comment]
https://github.com/ibus/ibus/assets/2330175/d3220db0-838d-43cc-a10b-82e0606ea7e1

**Comment by mike-fabian** [comment]
@itt533: Are you doing this using ibus-m17n or ibus-typing-booster?

**Comment by mike-fabian** [comment]
@itt533 Apparently you are using ibus-m17n.

**Comment by fujiwarat** [comment]
OK, I don't see any issue with  `m17n:sa:itrans`.
I'd ask you to take the screenshot of your step #5 only to show your issue clearly,

**Comment by mike-fabian** [comment]
Maybe it is easier to see in this video (high framerate 60 fps): 

https://github.com/ibus/ibus/assets/2330175/5a43d895-3453-4cdf-8838-008b2109dfe7

I am typing `bhA` using ibus-m17n with sa-itrans.
Each time the preedit gets an update, the cursor jumps to the left of the preedit, the updated preedit is displayed and the cursor jumps to the right again. This looks like a very quick  flicker. Depending on which system I test on, this is sometimes very hard to see and on some systems I cannot see it at all. For example I cannot see the flicker in a Fedora 38 virtual machine but I can see the flicker on a real Fedora 38 machine.

I tried whether other input methods show such a flicker when the preedit is updated and could not find any. As far as I can tell,

- ibus-typing-booster
- ibus-hangul
- ibus-anthy
- ibus-libpinyin

never show any flicker when the preedit is updated.

Maybe ibus-m17n updates the preedit in an unusual way?

Anyway, it doesn't look like a very serious problem to me and apparently one can avoid it by using ibus-typing-booster instead of ibus-m17n.

**Comment by fujiwarat** [comment]
> Each time the preedit gets an update, the cursor jumps to the left of the preedit, the updated preedit is displayed and the cursor jumps to the right again. This looks like a very quick flicker

@mike-fabian Thank you for the explanation. 
Seems you mean the the problem is the cursor position only but the output is no problem.

I guess it would be a small bug of the miss usage of IBus APIs in ibus-m17n.

**Comment by itt533** [comment]
> Maybe it is easier to see in this video (high framerate 60 fps):
> Peek.2023-07-21.15-52.mp4
> 
> I am typing `bhA` using ibus-m17n with sa-itrans. Each time the preedit gets an update, the cursor jumps to the left of the preedit, the updated preedit is displayed and the cursor jumps to the right again. 
[ ... ]
> Maybe ibus-m17n updates the preedit in an unusual way?
> 
> Anyway, it doesn't look like a very serious problem to me and apparently one can avoid it by using ibus-typing-booster instead of ibus-m17n.

Yes it's as you explained it. I did my video with 60 fps, (i used simplescreenrecorder).  With 30, the flicker was not noticeable.

Yes i just noticed that with ibus-typing-booster it's fine. With the latter only recently i have been able to shift the input method (previous_input_method and next_input_method). Ctrl+Up and Ctrl+Down respectively don't seem to work, but with custom ones (Ctrl+space and Shift+Ctrl+space) it seems to work. Still i need to see what is selected. Rocket not changing, and no contextual menu.

**Comment by mike-fabian** [comment]
> Yes i just noticed that with ibus-typing-booster it's fine. With the latter only recently i have been able to shift the input method (previous_input_method and next_input_method). Ctrl+Up and Ctrl+Down respectively don't seem to work, but with custom ones (Ctrl+space and Shift+Ctrl+space) it seems to work. Still i need to see what is selected. Rocket not changing, and no contextual menu.

Very surprising that `Ctrl+Up` and `Ctrl+Down` don’t work for you. I am not sure whether it is worth investigating though as you have found keys which work for you.

For seeing what currently is selected:

This is easy to do *unless* you use Gnome. If you are using any other desktop than Gnome, you can use `ibus-setup` and choose the option


Show property panel: [Always]

You can move this property panel to a place you like with the mouse.

When ibus-typing-booster is active, the property panel shows which dictionary and which input method currently have highest priority. Like in this screenshot:

![Screenshot](https://github.com/ibus/ibus/assets/2330175/8452cc43-eaab-4b4b-b91f-d2184199b6aa)

**Comment by fujiwarat** [comment]
Since it's just a bug of ibus-m17n, it's good to fix it.

**Comment by mike-fabian** [comment]
> The bug title is very bad to understand his issue, modified.

Thank you!

> Since it's just a bug of ibus-m17n, it's good to fix it.

Yes, I agree. I have spent a few hours on this today, but no success so far.

**Comment by mike-fabian** [comment]
While debugging, I noticed that `ibus_m17n_engine_update_preedit()` is almost always called twice for each update to the preedit, so I thought that maybe drawing the preedit twice causes the problem and I added code not to call ` ibus_engine_update_preedit_text_with_mode()` again when the contents of the preedit didn’t change since the last call. Like this:

```
diff --git a/src/engine.c b/src/engine.c
index d02e566..b399280 100644
--- a/src/engine.c
+++ b/src/engine.c
@@ -546,27 +546,37 @@ ibus_m17n_engine_destroy (IBusM17NEngine *m17n)
 static void
 ibus_m17n_engine_update_preedit (IBusM17NEngine *m17n)
 {
+    static IBusText *last_text = NULL;
+    static int last_cursor_pos = 0;
     IBusText *text;
     gchar *buf;
     IBusM17NEngineClass *klass = (IBusM17NEngineClass *) G_OBJECT_GET_CLASS (m17n);
 
+    if (!m17n)
+        return;
     buf = ibus_m17n_mtext_to_utf8 (m17n->context->preedit);
-    if (buf) {
-        text = ibus_text_new_from_string (buf);
-        if (klass->preedit_foreground != INVALID_COLOR)
-            ibus_text_append_attribute (text, IBUS_ATTR_TYPE_FOREGROUND,
-                                        klass->preedit_foreground, 0, -1);
-        if (klass->preedit_background != INVALID_COLOR)
-            ibus_text_append_attribute (text, IBUS_ATTR_TYPE_BACKGROUND,
-                                        klass->preedit_background, 0, -1);
-        ibus_text_append_attribute (text, IBUS_ATTR_TYPE_UNDERLINE,
-                                    klass->preedit_underline, 0, -1);
-        ibus_engine_update_preedit_text_with_mode ((IBusEngine *) m17n,
-                                                   text,
-                                                   m17n->context->cursor_pos,
-                                                   mtext_len (m17n->context->preedit) > 0,
-                                                   klass->preedit_focus_mode);
+    if (!buf)
+        return;
+    if (last_text && ibus_text_get_text(last_text) && strcmp(buf, ibus_text_get_text(last_text)) == 0 && m17n->context->cursor_pos == last_cursor_pos) {
+        g_free (buf);
+        return;
     }
+    text = ibus_text_new_from_string (buf);
+    last_text = ibus_text_new_from_string (buf);
+    last_cursor_pos = m17n->context->cursor_pos;
+    if (klass->preedit_foreground != INVALID_COLOR)
+        ibus_text_append_attribute (text, IBUS_ATTR_TYPE_FOREGROUND,
+                                    klass->preedit_foreground, 0, -1);
+    if (klass->preedit_background != INVALID_COLOR)
+        ibus_text_append_attribute (text, IBUS_ATTR_TYPE_BACKGROUND,
+                                    klass->preedit_background, 0, -1);
+    ibus_text_append_attribute (text, IBUS_ATTR_TYPE_UNDERLINE,
+                                klass->preedit_underline, 0, -1);
+    ibus_engine_update_preedit_text_with_mode ((IBusEngine *) m17n,
+                                               text,
+                                               m17n->context->cursor_pos,
+                                               mtext_len (m17n->context->preedit) > 0,
+                                               klass->preedit_focus_mode);
     g_free (buf);
 }
 
```

Unfortunately that didn’t improve anything at all, the behaviour is still exactly the same as before.

Still searching for other possible reasons ...

**Comment by mike-fabian** [comment]
Also, not only did the above attempt to draw the preedit only once when it is not changed not help to reduce the flicker, it also interferes with the compose support. 

That gave me the idea that the flicker appears because of the compose support. So I commented out these lines in `ibus_m17n_engine_process_key_event()` at:

https://github.com/ibus/ibus-m17n/blob/main/src/engine.c#L735

```C
    if (IBUS_ENGINE_CLASS (parent_class)->process_key_event (engine, keyval, keycode, modifiers)) {
      if (mtext_len (m17n->context->preedit) > 0) {
	gchar *buf;
	buf = ibus_m17n_mtext_to_utf8 (m17n->context->preedit);
	if (buf) {
	  IBusText *text;
	  text = ibus_text_new_from_string (buf);
	  ibus_engine_commit_text (engine, text);
	  g_free (buf);
	}
	minput_reset_ic (m17n->context);
      }
      return TRUE;
    }
```

and indeed this removes the flicker. But it also removes the compose support.

**Comment by mike-fabian** [comment]
Here are two videos comparing the flicker with ibus-m17n-1.4.19 and ibus-m17n-1.4.20. To make the flicker even more obvious, I did set a background colour and foreground colour for the preedit in the setup tool. 

In both videos I use `sa-itrans` and type `bhhhhhhhhhh` and then delete everything again by typing `backspace` many times.

Only ibus-m17n-1.4.19 produces flickering of the preedit, with ibus-m17n-1.4.20 the preedit does **not** flicker.

Video with ibus-m17n-1.4.19 (**without** the fix), considerable flicker:

https://github.com/ibus/ibus/assets/2330175/9f857a14-a8a0-4678-8ac2-0e4bd615a146

Video with ibus-m17n-1.4.20 (**with** the fix), no flicker:

https://github.com/ibus/ibus/assets/2330175/ead74ecd-ce59-4d5e-94e3-9f95a520112a

**Comment by mike-fabian** [comment]
Another test on Gnome Wayland. On Gnome Wayland, my change reduces the flicker less well then on Xorg, but it still reduces the flicker a lot. Weird things happen with the underlining in Gnome Wayland. The underlining is sometimes too short and sometimes too long when backspacing. But this weird problem with the underlining on Gnome Wayland happens both with and without my change, so it is probably unrelated to my change:

https://github.com/ibus/ibus/assets/2330175/8bf4cf76-735d-4f9e-a87e-e0a24ca3f393

**Comment by mike-fabian** [comment]
Fix is included in the ibus-m17n-1.4.20 release:
https://github.com/ibus/ibus-m17n/releases/tag/1.4.20

**Comment by itt533** [comment]
why did you close it?

**Comment by mike-fabian** [comment]
I cannot download that file. Are you sure that you tested the correct release? For me it made a significant difference as you can see in the videos above.

A perfect fix needs  new API in ibus. 

```
IBUS_ENGINE_CLASS (parent_class)->process_key_event (engine, keyval, keycode, modifiers)
```

hides the current preedit and then ibus-m17n shows it again.

For me adding a `ibus_engine_show_preedit_text ((IBusEngine *)m17n);` before and after calling the process_key_event() method of the parent class made the flicker mostly go away:

 https://github.com/ibus/ibus-m17n/commit/b74c6a68453e5a8132c7f2c104c446c8e5604963

A perfect fix would need a change in ibus to offer an extra parameter in the `process_key_event()` method of IbusEngineSimple so I could call it like this:

```
hide_preedit = mtext_len (m17n->context->preedit) == 0
IBUS_ENGINE_CLASS (parent_class)->process_key_event (engine, keyval, keycode, modifiers, hide_preedit)
```

or something similar.

**Comment by itt533** [comment]
> I cannot download that file. Are you sure that you tested the correct release? For me it made a significant difference as you can see in the videos above.

i upgraded to version 1.4.20
I re-uploaded the file, i got yhe same ink and i can download:
http://0x0.st/HLW0.bin

rename extension to .tar

**Comment by mike-fabian** [comment]
This time I can download it. But that site still gives me a warning in firefox "File not downloaded. Potential security risk.". So I have to download it on the commandline with wget, rename to .tar.gz, unpack. In future, to make it  easier, can you please attach videos here directly using one of the supported formats? (MOV, MP4, WEBM should work).

**Comment by mike-fabian** [comment]
Here is  your video as mp4s: ibus-1.4.20_flickering.mp4
https://github.com/ibus/ibus/assets/2330175/f9f5e025-5998-433b-b8f8-cc1706b99848

**Comment by mike-fabian** [comment]
Yes, sometimes there is still some flicker, but by direct comparision of ibus-m17n-1.4.19 and ibus-m17n-1.4.20 I found that it is considerably reduced.  In https://github.com/ibus/ibus/issues/2536#issuecomment-1662553935 I compared ibus-m17n-1.4.19 and ibus-m17n-1.4.20 on the same system and the updated version flickered less.

I cannot do a better fix at the moment without destroying compose support. 

As I explained in https://github.com/ibus/ibus/issues/2536#issuecomment-1686103616 the `process_key_event ()` method of the parent class IBusEngineSimple checks if the currently typed key is part of a compose sequence and if yes. it handles that key and returns TRUE. If the currently typed key  is **not** part of a compose sequence, the  `process_key_event ()` method of the parent class IBusEngineSimple  hides the preedit and returns FALSE. Then ibus-m17n needs to continue processing the key event and needs to show the hidden preedit again. Hiding and then showing the preedit again may cause this flicker. I put two calls to 

```
ibus_engine_show_preedit_text ((IBusEngine *)m17n);
```

directly around calling the `process_key_event ()` method  of the parent class, that reduced the flicker a lot for me. 
But it doesn't go away 100%. A 100% fix needs an improvement in ibus. To fix it, ibus could do one of these two things:

1. ibus **always** stops hiding the preedit when the currently typed key is not part of a compose sequence and the preedit is currently not empty
2. ibus adds some option to calling the `process_key_event ()` method of the parent class IBusEngineSimple and if that option is set, the preedit will not be hidden. Then ibus-m17n could request the preedit not to be hidden by setting that option.

If 1. is possible, I think 1. is the better way of doing it. I don’t really understand why IBusEngineSimple hides the preedit if the currently typed key is not in a compose sequence. Why not leave the preedit alone in that case? At the moment I see no reason why it should be hidden. But maybe I don’t understand it completely and there are situations where the preedit needs to be hidden in that case? If we cannot stop hiding a non-empty preedit **always** in the `process_key_event ()` method of the parent class IBusEngineSimple, then we need to do 2., i.e. add an option for the input engine inheriting from IBusEngineSimple to request not hiding the preedit.

So to make the flicker go away completely, some fix or improvement is needed in ibus, I did all I could to reduce the flicker as much as possible in ibus-m17n, I cannot do more in ibus-m17n without making the compose support worse. 

If I support compose **only** when the preedit is empty, then there would be no flicker of a non-empty preedit. But that would not be natural, then compose would sometimes seem to work (when the preedit is empty) and sometimes not (when there is something in the preedit). If the preedit is not empty, one would need to commit it (for example by typing a space), then remove that space by backspacing and then use the compose sequence. That is not nice at all. With the current code, a compose sequence started while the preedit is not empty automatically commits that preedit and then the compose sequence is handled. That is more natural and I want to keep it that way. 

So I cannot do more at the moment in ibus-m17n, a better solution needs an improvement in ibus.

If the flicker disturbs you so much, why not use ibus-typing-booster instead of ibus-m17n? ibus-typing-booster also has the option

☑️ Convert language specific digits to ASCII digits

which I implemented for you, so why not use that? Is there anything wrong with using typing booster?

**Comment by itt533** [comment]
I know about the option you implemented in ibus-typing-booster. I went back to ibus, it's for its ability (though very unstable) to remember input method for each application. Also ITB inserts automatically `0x200d` characters, but i'm not sure at what moment. Because of the pre-edit many hits are needed to move the cursor by one step, the appearing cursor is not the actual one, ... just check the my new screencast. 
[Ibus_typing_booster_left_right_cursor_navigation_-_2023-08-28_18.21.30_extract.webm](https://github.com/ibus/ibus/assets/90255020/a1a1826d-809d-4398-b113-77cb1d508289)

**Comment by mike-fabian** [comment]
> I went back to ibus, it's for its ability (though very unstable) to remember input method for each application.

Yes, I think that doesn’t always work well.

ibus-typing-booster cannot remember the input method for each application, but it is possible to choose the defaults of almost all settings of typing booster depending on the application.

For example you could choose to use `hi-itrans,te-itrans` in gedit and `te-itrans,hi-itrans` and switching the input off in gnome-terminal by default.

Whe an application gets focus, these "autosettings" depending on the application are applied. Within that application, one  can still use the key binding to change the settings again. Wehn the application loses focus, the settings are reverted to the state before it got focus.

This "autosettings" feature works very reliable on Xorg, it is a bit more problematic on Wayland. Usually it works on Wayland as well but sometimes it suddenly stops working.

The autosettings feature is described here:

https://mike-fabian.github.io/ibus-typing-booster/docs/user/#2_2_7

**Comment by mike-fabian** [comment]
> Because of the pre-edit many hits are needed to move the cursor by one step, the appearing cursor is not the actual one, ... just check the my new screencast.

I see two things in your video:

First:


Sometimes two cursors are displayed, a blue one and a black one with little triangles at the bottom. This is caused by a bug in Pango which is still not fixed, see:

https://gitlab.gnome.org/GNOME/pango/-/merge_requests/653#note_1795767

This problem can happen with ibus-m17n as well.

Second:

You correctly note that sometimes multiple steps are needed to step over one character in the preedit if that character was composed by typing several characters.

For example, when typing `shata` with `te-itrans` as in you video, the final result is శత both in ibus-m17n and in ibus-typing-booster. But ibus-m17n commits the `sha -> శ` already as soon as the following `t` is typed. Whereas typing booster keeps the whole word in preedit until the next commit key (usually a space) to be able to predict completions of that word.

You can see the preedit indicated by the underline in typing booster. It is good to see what happens if you turn on that underline as well in `te-itrans (m17n)`. You see then that `te-itrans (m17n)` commits much more often than ibus-typing-booster and never keeps a long string in preedit. Also you cannot even use arrow left `Left` to go back in the preedit in `te-itrans (m17n)`: If you type `sh` + `Left` in  `te-itrans (m17n)` then the preedit containing శ్ is committed. You see that because the underline disappears as soon as the arrow left is typed. But when you use `backspace` instead of `Left` you see that `sh` + `backspace` deletes the part typed by the `h` and leaves only the part typed by the `s`, i.e. స్.

In typing booster however, typing `sh` + `Left` puts the cursor actually in between the `s` and the `h` in the preedit, but it is impossible to display a cursor at that position. So the cursor is still displayed **after** the `sh` even though it is logically between the typed keys `s` and the `h`. If one types `Left` once more, then the cursor is to the left of the `sh` where it can be displayed again at the correct position.

As soon as something is committed, there is no difference anymore between ibus-m17n and ibus-typing-booster. If you type `shata ` (space at the end), you get "శత " both with ibus-m17n **and** with ibus-typing-booster and now moving backwards over this with `Left` needs 3 times arrow left no matter whether this was written using ibus-m17n or ibus-typing-booster.

So ibus-typing-booster lets you edit a whole word in preedit by going back within that word with `Left` which needs one key press for each input character typed. So it would actually be possible to insert something between a typed `sh`, you would type `sh` + `Left` + a + `arrow-right` + `space` to get "సహ్ ". That particular example is not very useful, I know, I am just trying to explain how it works.

One thing you can do if you never ever want to edit preedits by going backwards to somewhere inside the preedit using `Left`, you could define `Left` as an additional commit key. Like this:

![Screenshot](https://github.com/ibus/ibus/assets/2330175/7d53a48e-b7f1-4e63-a115-dfeb9abbea9c)

(The default for that setting is empty).

If you do this, the first `Left` will immediately commit the current preedit, just as it does in ibus-m17n and from then on stepping left over the committed characters works the same.

There is still a tiny difference when doing this:

In ibus-m17n, `Left` commits and an arrow-left event is sent to the application, i.e. if you type `a` + `Left` in ibus-m17n, you get "|అ" (where | indicates the cursor position) but in ibus-typing-booster you get "అ|" (where | indicates the cursor position). All subsequent presses of `Left` will behave the same because there is no preedit anymore.

Theoretically I could add another configurable command in ibus-typing-booster, something like `commit_and_forward_commit_key` which when set to `Left` would commit **and** forward/pass the key which triggered to the commit to the application. That would mimic the behaviour of ibus-m17n even more closely.

**Comment by mike-fabian** [comment]
Do you think that setting the `commit` command to `Left` is an improvement for you?
Would you like to have a new command `commit_and_forward_commit_key` to be able to mimic the behaviour of ibus-m17n even closer?

**Comment by itt533** [comment]
pre-edit might be useful for `ITB` as it looks at what is coming and suggests from the dictionary. I'd prefer having no pre-edit at all with characters dumped flawlessly. So far, the best input method i know is that of `emacs`. Hit `C-x m \` and `telugu-itrans`. The ability to navigate along the string, 1 by 1 character from `utf` point of view is nice.  Emacs jumps over ligatured characters. That's a limitation.

As for the screencast i have seen that blue cursor. Do you have the ability to know how much cpu resource is taking each input with `Ibus` or `ITB`?

As for the ability to use input methods based on app, I'd prefer the `Ibus` way rather than preset associations in config. But `Ibus` as a problem (i have reported it) which is that simply activating a window is not waking up `Ibus`, I have to hit a key and then only the right input method is set.

**Comment by fujiwarat** [comment]
I think this bug report is not good. Please don't mention for engine specific issues here and change the title before I evaluate it.

**Comment by fujiwarat** [comment]
I will evaluate this can be reopen later.

**Comment by mike-fabian** [comment]
> pre-edit might be useful for `ITB` as it looks at what is coming and suggests from the dictionary. I'd prefer having no pre-edit at all with characters dumped flawlessly. So far, the best input method i know is that of `emacs`. Hit `C-x m \` and `telugu-itrans`. The ability to navigate along the string, 1 by 1 character from `utf` point of view is nice.

No preedit at all is not possible. Even `telugu-itrans` in emacs uses a preedit in the same way ibus-m17n does. See the video below.

I am typing `shata`. While I am typing `sha`, it is underlined, because it is in preedit. It has to be because that character is not yet finished. When the `t` is typed, it is clear that the `sha` will not change anymore and that part can be committed. When the following `a` is typed, the part produced by the `ta` is underlined, i.e. in preedit. When the final space is typed, the `ta` part is committed.

Then I type `shata` again, but this time first typing `sha` then backspacing twice to remove the `a` and the `h`. Then typing `ah` again, typing `t` (commit of `sha` happens now), then `a`, then backspacing, then adding the `a` again, then space to commit.

Backspacing after typing `sha` (శ) to remove the `ha` and leave `s` (స్) is only possible because the శ was still in preedit. As soon as it is committed, one backspace deletes the whole శ.

That behaviour is the same in `te-itrans` in ibus-m17n and `telugu-itrans` in emacs. Both keep the preedit as long as changes to the current character are still possible. If one character is finished, it is committed. So there  is no difference here. Only ibus-typing-booster is different, it keeps the preedit for longer than one character to be able to do predictions.


https://github.com/ibus/ibus/assets/2330175/f5f57a51-9288-4b6e-986c-2e7760dc0c8c

**Comment by itt533** [comment]
> No preedit at all is not possible. Even `telugu-itrans` in emacs uses a preedit in the same way ibus-m17n does. See the video below.

Yes emacs uses pre-edit. It is for ligatures. I was only saying that it is the best that i came across. It is in pre-edition mode as long as the being composed character can be further complexified. It is related to the grammar.
 
> That behaviour is the same in `te-itrans` in ibus-m17n and `telugu-itrans` in emacs. Both keep the preedit as long as changes to the current character are still possible. If one character is finished, it is committed. So there is no difference here. 

Not exactly the same... after typing `a` in `sha` ibus -m17n is still in pre-edition but backspace deletes the whole character.

Isn't there any way to get rid of Pango and develop the code from scratch in c?

**Comment by mike-fabian** [comment]
> > No preedit at all is not possible. Even `telugu-itrans` in emacs uses a preedit in the same way ibus-m17n does. See the video below.
> 
> Yes emacs uses pre-edit. It is for ligatures. I was only saying that it is the best that i came across. It is in pre-edition mode as long as the being composed character can be further complexified. It is related to the grammar.

Yes, that is what I was trying to say. Both ibus-m17n and emacs keep things as long in preedit as necessary, as long as the composed character is not finished (can be further complexified). If that is not the case anymore, they both commit.

> > That behaviour is the same in `te-itrans` in ibus-m17n and `telugu-itrans` in emacs. Both keep the preedit as long as changes to the current character are still possible. If one character is finished, it is committed. So there is no difference here.
> 
> Not exactly the same... after typing `a` in `sha` ibus -m17n is still in pre-edition but backspace deletes the whole character.

Yes, you are right, there is a tiny difference here. ibus-m17n already "finishes" the `sha` to శ but does not commit yet and keep it in the preedit. It is sort of "committed to preedit". So you can delete it already with one backspace. The real commit happens when the next character, for example space is typed. In Emacs, the శ produced by `sha` is still kept in its "unfinished" state so you can still delete it part by part using 3 times backspace. I.e. emacs does not "commit to preedit" but keeps the unfinished thing in preedit. But then  when the next charater, for example space is typed, emacs commits as well.

So there is a very tiny difference here, but not much.

I thought you wanted a commit as early as possible. Then you should prefer the ibus-m17n behaviour, which commits the `sha` to the finished state a bit earlier than emacs.

The behaviour of emacs is actually somewhere in between ibus-m17n and ibus-typing-booster, emacs keeps the unfinished stuff a bit longer than ibus-m17n, but not as long as ibus-typing-boosster. ibus-typing-booster keeps the "unfinished" stuff until the end of the whole word, i.e. until the whole word is finished, you can still delete all the individual keys typed individually and you need as many `Left` events to step left through the word in preedit as there are characters typed, you could still change details somewhere in that unfinished stuff. Although that is a bit difficult as you would need to count steps because, for example, the cursor cannot be displayed somewhere "inside" an శ, only before the శ or after the శ.

**Comment by mike-fabian** [comment]
> Do you think that setting the `commit` command to `Left` is an improvement for you? Would you like to have a new command `commit_and_forward_commit_key` to be able to mimic the behaviour of ibus-m17n even closer? If you want such a `commit_and_forward_commit_key` command, please open an issue at:
> 
> https://github.com/mike-fabian/ibus-typing-booster/issues

Did you try that? Does it improve things for you?

**Comment by itt533** [comment]
> Did you try that? Does it improve things for you?
no, i haven't because of what i have already mentioned with `ITB`. Actually it could be interesting in Ibus to commit without moving the cursor back by one step. Up to not so long ago i had only used scim. As far as i remember whatever first hit was instantly committed. Actually the problem with Ibus is that because of the pre-edit, in goldendict I was no more able to see a list of words after hitting the first key. I have to issue either `space`, which of course i have delete if i further input, or issue `left` which i would have to compensate with `right` if further input. The problem with scim was that support was dropped. Support for qt5 was not added.

**Comment by mike-fabian** [comment]
> > Did you try that? Does it improve things for you?
> 
> no, i haven't because of what i have already mentioned with `ITB`.

But I told you how this can be improved.

You wrote:

> Also ITB inserts automatically 0x200d characters

I never saw that and I don’t think it happens. If it really happens, tell me when it happens.

> Because of the pre-edit many hits are needed to move the cursor by one step,

This problem goes away if you assign `Left` as a key binding for the command `commit`, then moving the cursor left commits and you don’t need so many hits anymore to move through the text.

> Actually it could be interesting in Ibus to commit without moving the cursor back by one step.

In ibus-typing-booster, you can bind as many keys as you like to the `commit` command. In addition to `Left` you could also use other keys.

The command `commit` just commits and does **nothing** else, no cursor movement, nothing else.

I was suggesting adding an additional command `commit_and_forward_commit_key` to be able to commit **and** send the key which was used to commit. When binding that to `Left`, then `Left` would do the same as it does in ibus-m17n, commit and move the cursor left. 

> Up to not so long ago i had only used scim. As far as i remember whatever first hit was instantly committed. Actually the problem with Ibus is that because of the pre-edit, in goldendict I was no more able to see a list of words after hitting the first key. I have to issue either `space`, which of course i have delete if i further input, or issue `left` which i would have to compensate with `right` if further input. The problem with scim was that support was dropped. Support for qt5 was not added.

But scim of course also used preedit. You have to use preedit, because a commit happens after each key, then `sha` with `te-itrans would give you స్హ్అ instead of శ. When you type `s` you see స్, if you commit this and then type `h` you would just add a హ్ after that getting స్హ్ as the final result, not శ్. So one has to keep the character in preedit until it is finished.

As you wrote: “It is in pre-edition mode as long as the being composed character can be further complexified. It is related to the grammar.”

scim also did that.

Theoretically it is sometimes possible to commit but use a feature called surrounding text to get the commit back if needed, i.e. if `s` is typed and స్ is committed and then `h` is typed, one can ask the application what is left of the cursor and if there is స్ then fetch it from the application, delete it there, combine it with the `h` to get `sh` శ్ and commit that again. That would require perfectly working surrounding text. But surrounding text is very buggy and there are many applications where it doesn’t work at all. So usually one has to use preedit at least  “as long as the being composed character can be further complexified.” scim did that, ibus-m17n does that, emacs does that as well.

So I would encourage you to try ibus-typing-booster again and bind `Left` to the `commit` command. I am pretty sure that improves things for you.

**Comment by itt533** [comment]
> > Because of the pre-edit many hits are needed to move the cursor by one step,
> 
> This problem goes away if you assign `Left` as a key binding for the command `commit`

i tried it, it solves some problems but it also creates a new problem.  If one types `sh` and commits, he will have to delete the `h` and reproduce it and then only be able to add `a` to make `శ`. Otherwise he will have `శ్అ`.

> ... would do the same as it does in ibus-m17n, commit and move the cursor left.

I'm not praising ibus-m17n. The trigger for this bug report was a problem in ibus, and the other one  i reported is really annoying and reached the state of being ignored.

> But scim of course also used preedit. You have to use preedit, because a commit happens after each key, then `sha` with `te-itrans would give you స్హ్అ instead of శ. When you type `s`you see స్, if you commit this and then type`h` you would just add a హ్ after that getting స్హ్ as the final result, not శ్. So one has to keep the character in preedit until it is finished.

you must be right, now I remember, i was using inscript input methods. All consonants are input by default combined with a vowel, which allows to be immediately committed. This way, goldendict could display a list of words at first hit.

> So I would encourage you to try ibus-typing-booster again and bind `Left` to the `commit` command. I am pretty sure that improves things for you.

you did not remind the problem of input method per app. Also that bug of Pango as you said. Here in the beginning i said that assigning `left` to commit is not satisfactory. Ibus is the right basis but it looks like dirty code.

**Comment by mike-fabian** [comment]
> > > Because of the pre-edit many hits are needed to move the cursor by one step,
> > 
> > 
> > This problem goes away if you assign `Left` as a key binding for the command `commit`
> 
> i tried it, it solves some problems but it also creates a new problem. If one types `sh` and commits, he will have to delete the `h` and reproduce it and then only be able to add `a` to make `శ`. Otherwise he will have `శ్అ`.

Yes, of course. If you commit, you cannot edit committed stuff easily anymore. But that is always like that, no matter whether you use ibus-m17n, ibus-typing-booster, scim, emacs, ...

You said you didn’t like that ibus-typing-booster keeps a long preedit because then stepping back with `Left` needs many steps. In ibus-m17n, `Left` commits and steps back.
In emacs, `Left` after typing `sh` (శ్) just does nothing, the శ్ stays in preedit, nothing at all happens, not particularly useful.

I think making go `Left` back into the preedit to edit somewhere **inside** the preedit (which is what ibus-typing-booster does by default) can be useful, although it is difficult to understand especially when complex scripts like Telugu are involved.

What ibus-m17n does, i.e. making `Left` commit and step back is also useful.

I have already implemented a new command `commit_and_forward_key` which can be bound to `Left` if desired, see this:

https://github.com/mike-fabian/ibus-typing-booster/issues/461

It will be in the next release.

Using that, `Left` commits and steps back just as in ibus-m17n. What else should one do? Doing nothing at all like in Emacs is sure not any better.

So I don't see at all what new problem this create. If a commit happens, the keys typed which composed the text which was committed are gone, so you cannot edit individual typed keys anymore. That is always the case if you commit something like `sh` శ్. You cannot easily add an `a` anymore to make this `sha` శ. This cannot be fixed **except** using surrounding text: With surrounding text, the input method could try to get the `sh` శ్ back into preedit, then add the `a` as well to the preedit and then commit `sha` శ. But none of the input methods we are currently talking about does this. Neither ibus-typing-booster nor emacs nor ibus-m17n. Theoretically it is possible, but there are so many applications where surrounding text does not work at all or is extremely buggy, that we would run into a many other problems when trying the route of using surrounding text.

Even telugu-itrans in Emacs does not use surrounding text, even though in Emacs it would be far easier to implement than in ibus. ibus has to work with many applications which may or may not support surrounding text correctly. telugu-itrans in Emacs however has to work only in Emacs and in Emacs it should be possible to get the text left of the cursor when you type `a` and if that is `sh` then put it together again to `sha`. But even Emacs does **not** do that currently for telugu-itrans.

> > ... would do the same as it does in ibus-m17n, commit and move the cursor left.
> 
> I'm not praising ibus-m17n. The trigger for this bug report was a problem in ibus, and the other one i reported is really annoying and reached the state of being ignored.

Which bug is ignored? Are you talking about this but here? The flicker? It is not being ignored, I improved it already and eventually we will get a perfect solution, but this needs more time.

> > But scim of course also used preedit. You have to use preedit, because a commit happens after each key, then `sha` with `te-itrans would give you స్హ్అ instead of శ. When you type `s`you see స్, if you commit this and then type`h` you would just add a హ్ after that getting స్హ్ as the final result, not శ్. So one has to keep the character in preedit until it is finished.
> 
> you must be right, now I remember, i was using inscript input methods. All consonants are input by default combined with a vowel, which allows to be immediately committed. This way, goldendict could display a list of words at first hit.

Yes, inscript is very different from itrans, in inscript every single keypress can commit already.

> > So I would encourage you to try ibus-typing-booster again and bind `Left` to the `commit` command. I am pretty sure that improves things for you.
> 
> you did not remind the problem of input method per app. Also that bug of Pango as you said. Here in the beginning i said that assigning `left` to commit is not satisfactory. Ibus is the right basis but it looks like dirty code.

That the setting in ibus to remember the used input methods per app and not use one global input method for the whole desktop has some problems is not ignored either. I think Fujiwara San, the main ibus developer has it on his radar and I am sure it will eventually be fixed.

And the Pango bug will also be fixed soon, there are patches fixing the Pango bug already which will probably be included soon.

**Comment by mike-fabian** [comment]
> I have already implemented a new command commit_and_forward_key which can be bound to Left if desired, see this:
> 
> https://github.com/mike-fabian/ibus-typing-booster/issues/461
> 
> It will be in the next release.

This is in the ibus-typing-booster-2.24.0 release: https://github.com/mike-fabian/ibus-typing-booster/releases/tag/2.24.0

**Comment by fujiwarat** [comment]
I'm sorry but many ibus-typing-booster comments are not relative with this issue and confuse me.
Please file a new issue for ibus-typing-booster and copy the relative comments.
I will delete the unnecessary comments later.

**Comment by mike-fabian** [comment]
https://github.com/ibus/ibus-m17n/releases/tag/1.4.24 includes a fix now which together with ibus >= 1.5.29 fixes the preedit flicker perfectly.

This is the fix on the ibus-m17n side: https://github.com/ibus/ibus-m17n/commit/2331934b88f9d85f2cd74ab850131574f68a8b1c

The previous workaround is not needed anymore and was removed.

**Comment by fujiwarat** [comment]
Again, don't close this bug!

**Comment by itt533** [comment]
> https://github.com/ibus/ibus-m17n/releases/tag/1.4.24 includes a fix now which together with ibus >= 1.5.29 fixes the preedit flicker perfectly.
only this package needs an update?
can you show a screencast?

**Comment by mike-fabian** [comment]
> > https://github.com/ibus/ibus-m17n/releases/tag/1.4.24 includes a fix now which together with ibus >= 1.5.29 fixes the preedit flicker perfectly.
> 
> only this package needs an update? can you show a screencast?

No, only an update of this package is not enough (I guess this is why Fujiwara San reopened the bug after I closed it).

You need that update **and** an update of ibus including this branch:

https://github.com/fujiwarat/ibus/commit/7bfd2ad82014bdaffb39e59c166d1d7c507e4ee7

and including this pull request for the branch:

https://github.com/fujiwarat/ibus/pull/6

This is **not** yet in ibus-1.5.29~rc2-1.fc39.x86_64

So please wait a bit more.

**Comment by fujiwarat** [comment]
Sorry for the late response.
I evaluated this issue with https://github.com/ibus/ibus/pull/2566 again.
I will integrate the patch after Fedora 40 GA.

**Comment by fujiwarat** [comment]
The main patch is integrated but I found a regression and the referred another patch is under the testing in Fedora 40.
